documentation/high-level-introduction.md

```diff
@@ -0,0 +1,251 @@
+
+
+## XML Structure
+  - The top level of each project\target\prop file is a `<Project>` node which contains many children
+  - As noted above, each specific file type tends to contain certain items. However **these are all conventions**! 
+  - If the XML is invalid (unclosed tag, etc) your build may fail in strange ways.
+    - macios has an xml check as pat of `make msbuild`
+
+**Logic Constructs**
+
+  - Many nodes can have Conditions which when false will prevent them and their children from being used in a given build.
+    - Conditions can even prevent entire files from being included. 
+    - Almost everything can have one and they are very powerful
+  - There are XML constructs that mirror switch statements, recursion, and many other programming elements.
+  - As everything is strings, comparisions involves a number of quotes:
+
+            <SwiftVerbose Condition="'$(SwiftVerbose)' == ''">false</SwiftVerbose>
+
+    - Outer `"` for the contents of the condition
+    - `'` around the variable itself
+    - `'’` for empty string
+  - Remember, unset variables are valid to access and provide empty string
+
+**List Transforms**
+
+  - If your contents contain a `%` then you will be run like a “for each” for each element in the relevant itemgroup.
+  - See the [XA doc](https://github.com/xamarin/xamarin-android/blob/master/Documentation/guides/MSBuildBestPractices.md#item-group-transforms) and [msbuild docs](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-transforms?view=vs-2017) for more details. It’s powerful but complex.
+
+
+## MSBuild Tasks
+- These are C# types defined in a library. 
+- The specifics of creating a new task library is outside the scope of this guide.
+- There are two common base classes for Tasks:
+  - `Task` - A generic “I want to do some work” base class with and `Execute ()` method to override
+  - `ToolTask` - A base class that makes invoking a single external tool significantly easier
+- Parameters are passed via public properties defined on the task
+
+```
+public abstract class CodesignNativeLibraries : Task 
+{
+    public string SessionId { get; set; }
+    
+    [Required]
+    public string AppBundleDir { get; set; }
+}
+```
+
+  - The can be marked [Required] which will fail the build if not given a non empty string value
+  - Strings are the common type, but you can use types like bool and msbuild will convert for you
+  - ITaskItem[] are need to pass Items (lists) to \ from tasks when the task may need to modify them
+- They are later invoked with:
+
+```
+     <UsingTask TaskName="Xamarin.iOS.Tasks.CodesignNativeLibraries" AssemblyFile="Xamarin.iOS.Tasks.dll" />
+     
+    <Target Name="_CodesignNativeLibraries">
+                    <CodesignNativeLibraries
+                            SessionId="$(BuildSessionId)"
+                            AppBundleDir="$(AppBundleDir)"
+                            >
+                    </CodesignNativeLibraries>
+    </Target>
```

Comment:
Could this be a simpler example? Something that wouldn't require some thinking about Mac/iOS? I'm thinking if someone has no idea about those things.

Reply:
Made a more trivial example.